ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -549,6 +549,67 @@ public func concatMap<T, U, E>(transform: T -> SignalProducer<U, E>)(producer: S
     return producer |> map(transform) |> concat
 }
 
+/// Returns a signal that forwards values from the latest signal sent on
+/// `producer`, ignoring values sent on previous inner signals.
+///
+/// An error sent on `producer` or the latest inner signal will be sent on the
+/// returned signal.
+///
+/// The returned signal completes when `producer` and the latest inner signal
+/// have both completed.
+public func latest<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> SignalProducer<T, E> {
+	return SignalProducer<T, E> { sink, disposable in
+		let producerCompleted = Atomic(false)
+		let latestCompleted = Atomic(false)
+		
+		let completeIfNecessary: () -> () = {
+			if producerCompleted.value && latestCompleted.value {
+				sendCompleted(sink)
+			}
+		}
+
+		let latestDisposable = SerialDisposable()
+		disposable.addDisposable(latestDisposable)
+
+		producer.startWithSignal { signal, producerDisposable in
+			disposable.addDisposable(producerDisposable)
+			
+			signal.observe(
+				next: { signal in
+					latestCompleted.value = false
+					
+					signal.startWithSignal { signal, signalDisposable in
+						latestDisposable.innerDisposable = signalDisposable
+						
+						signal.observe(SinkOf { event in
+							switch event {
+							case .Completed:
+								latestCompleted.value = true
```

Comment:
Ugh, you're right. Checking that `signalDisposable.disposed` is false before setting `latestCompleted` should remedy that. However, that introduces a third synchronization object making this code a bit hairy. 

Alternatively, we could package the two completed bools and the latest signal in a `LatestState` struct. Wrap that in an `Atomic` so that we only have one lock to consider. Then in the inner completion handler check if it's still the latest signal and only set `latestCompleted` if so.